src/cargo/util/graph.rs

```diff
@@ -154,17 +389,333 @@ impl<N: fmt::Display + Eq + Ord + Clone, E: Clone> fmt::Debug for Graph<N, E> {
+    age: GraphAge,
+}
+
+#[test]
+fn demonstrate_stack_graph_can_read_all_clones() {
+    let mut graph = StackGraph::new();
+    let mut stack = Vec::new();
+    graph.add_edge(1, 2, 1);
+    stack.push(graph.clone());
+    graph.add_edge(2, 3, 2);
+    stack.push(graph.clone());
+    graph.add_edge(2, 3, 3);
+    stack.push(graph.clone());
+    // violate stack discipline, so a deep clone is needed
+    graph = stack[0].clone();
+    graph.add_edge(2, 3, 4);
+    assert_eq!(
+        stack.iter().map(|g| g.contains(&2)).collect::<Vec<bool>>(),
+        [false, true, true]
+    );
+    assert_eq!(stack[1].borrow().edge(&2, &3).collect::<Vec<_>>(), [&2]);
+    assert_eq!(stack[2].borrow().edge(&2, &3).collect::<Vec<_>>(), [&2, &3]);
+    assert_eq!(graph.borrow().edge(&2, &3).collect::<Vec<_>>(), [&4]);
+}
+
+impl<N: Clone, E: Clone> Clone for StackGraph<N, E> {
+    fn clone(&self) -> Self {
+        *RefCell::borrow_mut(&self.other_refs)
+            .entry(self.age)
+            .or_insert(0) += 1;
+        StackGraph {
+            inner: Rc::clone(&self.inner),
+            other_refs: Rc::clone(&self.other_refs),
+            age: self.age,
         }
     }
 }
+
+fn remove_from_other_refs(borrow: &mut RefMut<'_, BTreeMap<GraphAge, usize>>, age: GraphAge) {
+    if let std::collections::btree_map::Entry::Occupied(mut val) = borrow.entry(age) {
+        *val.get_mut() -= 1;
+        if val.get() == &0 {
+            val.remove();
+        }
+    }
+}
+
+impl<N: Clone, E: Clone> Drop for StackGraph<N, E> {
+    fn drop(&mut self) {
+        remove_from_other_refs(&mut RefCell::borrow_mut(&self.other_refs), self.age);
+    }
+}
+
+impl<N: Eq + Hash + Clone, E: Clone + PartialEq> StackGraph<N, E> {
+    pub fn new() -> StackGraph<N, E> {
+        let inner = Graph::new();
+        let age = inner.len();
+        let mut other_refs = BTreeMap::new();
+        other_refs.insert(age, 1);
+        StackGraph {
+            inner: Rc::new(RefCell::new(inner)),
+            other_refs: Rc::new(RefCell::new(other_refs)),
+            age,
+        }
+    }
+
+    pub fn borrow(&self) -> StackGraphView<'_, N, E> {
+        let inner = RefCell::borrow(&self.inner);
+        assert!(self.age.len_edges <= inner.len().len_edges, "The internal Graph was reset to something smaller before you tried to read from the StackGraphView");
+        assert!(self.age.len_nodes <= inner.len().len_nodes, "The internal Graph was reset to something smaller before you tried to read from the StackGraphView");
+        StackGraphView {
+            inner,
+            age: self.age,
+        }
+    }
+
+    /// Gets mutable access to the inner `Graph`. Uses `other_refs` to determine if a deep clone is
+    /// needed, and runs `reset_to` to get the `Graph` into the state it was in.
+    ///
+    /// It is the responsibility of the caller to set `self.age` and call `add_to_other_refs` after the new age is determined.
+    fn activate(&mut self) -> RefMut<'_, Graph<N, E>> {
```

Comment:
Why is this called `activate`? And could the comment explain why the caller has to maintain the object invariants.